src/fileimpl.cpp

```diff
@@ -551,27 +551,40 @@ class Grouping
     struct zim_MD5_CTX md5ctx;
     zim_MD5Init(&md5ctx);
 
+    unsigned char ch[CHUNK_SIZE];
     offset_type checksumPos = header.getChecksumPos();
-    offset_type currentPos = 0;
+    offset_type toRead = checksumPos;
+    
     for(auto part = zimFile->begin();
         part != zimFile->end();
         part++) {
       std::ifstream stream(part->second->filename(), std::ios_base::in|std::ios_base::binary);
 
-      char ch;
-      for(/*NOTHING*/ ; currentPos < checksumPos && stream.get(ch).good(); currentPos++) {
-        zim_MD5Update(&md5ctx, reinterpret_cast<const uint8_t*>(&ch), 1);
+      while(toRead>=CHUNK_SIZE && stream.read(reinterpret_cast<char*>(ch),CHUNK_SIZE).good()) {
+        zim_MD5Update(&md5ctx, ch, CHUNK_SIZE);
+        toRead-=stream.gcount();
       }
+
+      // It reads the remaining amount of file when we reach the end of the file
+      if(stream.good()){
```

Comment:
Please add a new comment here: 
```
Previous read was good, so we have exited the previous `while` because `toRead<CHUNK_SIZE`. Let's try to read `toRead` chars and process them later.
Else, the previous `while` exited because we didn't succeed to read `CHUNK_SIZE`, and we still have some data to process before changing part.